Allow no tabs to be selected in TabBar and TabContainer

kitbdev opened the pull request:
- closes https://github.com/godotengine/godot-proposals/issues/8233
- related #21223
- fixes #74653
- fixes #73785
- fixes `remove_tab()` and `move_tabs()` not updating previous tab properly.
- related: https://github.com/godotengine/godot/issues/50930#issuecomment-1782019598

When TabContainer children visibility changes, that tab is marked as current and other tabs are deselected.

Adds deselect_enabled property to TabBar and TabContainer so that no tabs can be selected (disabled by default).
current_tab can now be -1 if this is enabled or there are no valid tabs.

Considerations:
- If there is only one child of a TabContainer and deselect_enabled is false, then the child cannot be hidden.
- This also prevents multiple TabContainer children from being visible at once.
- Changing visibility of TabContainer children changes the current tab, so when editing you need to make sure to switch to the tab you want to be the first selected tab before saving the scene.

kitbdev committed:
Allow tab deselection

KoBeWi commented:
How does this fix https://github.com/godotengine/godot/issues/21223?

Also there is some bug when adding a first child to a newly created TabContainer. The child is invisible and you can't make it appear by changing tabs.

kitbdev commented:
Oh, I misread #21223, this only address the 'bonus' part of it, where clicking the eye to change visibility sets the current tab and the others are hidden. Changed to just be related.

And I think the invisible child issue is partly because I changed how the tab_changed signal fires in edge cases, but maybe I shouldn't change it for compatibility.
When adding the first child to a TabContainer, should the tab_changed signal be emitted?

AThousandShips reviewed (commented):
The change to the default means that users who previously selected the first tab will have no tabs selected, this might cause unexpected behaviour

KoBeWi commented on doc/classes/TabBar.xml:
```diff
@@ -229,8 +229,11 @@
+			The index of the current selected tab. A value of [code]-1[/code] means that no tab is selected.
+		</member>
+		<member name="deselect_enabled" type="bool" setter="set_deselect_enabled" getter="get_deselect_enabled" default="false">
+			If [code]true[/code], all tabs can be deselected so that no tab is selected. Click on the [member current_tab] to deselect it.
```
```suggestion
			If [code]true[/code], all tabs can be deselected so that no tab is selected. Click on the current tab to deselect it.
```
Referencing the property is confusing here.

KoBeWi commented on doc/classes/TabBar.xml:
```diff
@@ -229,8 +229,11 @@
-		<member name="current_tab" type="int" setter="set_current_tab" getter="get_current_tab" default="0">
-			Select tab at index [code]tab_idx[/code].
+		<member name="current_tab" type="int" setter="set_current_tab" getter="get_current_tab" default="-1">
+			The index of the current selected tab. A value of [code]-1[/code] means that no tab is selected.
```
Should mention that -1 can be set only with `deselect_enabled` active.

kitbdev replied:
It can be set to -1 if deselect is active or if no tabs can be selected, I'll mention it.

KoBeWi commented on scene/gui/tab_bar.cpp:
```diff
@@ -640,7 +644,19 @@ int TabBar::get_tab_count() const {
+	if (p_current == -1) {
+		// An index of -1 is only valid if deselecting is enabled or there are no valid tabs.
+		bool no_valid_tabs = true;
+		for (int i = 0; i < get_tab_count(); i++) {
```
This check is pointless when `deselect_enabled` is enabled.

KoBeWi commented on scene/gui/tab_bar.cpp:
```diff
@@ -640,7 +644,19 @@ int TabBar::get_tab_count() const {
+				break;
+			}
+		}
+		ERR_FAIL_COND_MSG(!deselect_enabled && !tabs.is_empty() && !no_valid_tabs, "Cannot deselect tabs, deselection is not enabled.");
```
```suggestion
		ERR_FAIL_COND_MSG(!deselect_enabled && !no_valid_tabs, "Cannot deselect tabs, deselection is not enabled.");
```
`no_valid_tabs` already handles it.

KoBeWi commented:
The logic for setting -1 can be improved. Pseudocode:
```
can_deselect = deselect_enabled
if (!can_deselect):
	for each tab:
		if tab.invisible or tab.disabled:
			can_deselect = true
			break

ERR_FAIL_COND_MSG(!can_deselect)
```

KoBeWi commented on scene/gui/tab_bar.cpp:
```diff
@@ -1520,6 +1550,9 @@ void TabBar::ensure_tab_visible(int p_idx) {
 	if (!is_inside_tree() || !buttons_visible) {
 		return;
 	}
+	if (p_idx == -1) {
```
Shouldn't this check for `deselect_enabled`?
Also might be worth mentioning it in the docs, as this is rather unexpected.

kitbdev replied:
I don't want it to error when -1 is passed whether `deselect_enabled` is true or false, since it is a value that current_tab can be.
I'm not sure what is unexpected about it, it wouldn't do anything otherwise but error.

KoBeWi replied:
> since it is a value that current_tab can be.

But it can be -1 only when `deselect_enabled` or there are no tabs, right? When -1 `current_tab` results in an error, so should this method I think.

kitbdev replied:
Ok, I'll have to have another check for no valid tabs then.

KoBeWi replied:
Maybe it can be a method, since you need it a few times. Something like `can_deselect()`.

KoBeWi commented on scene/gui/tab_bar.cpp:
```diff
@@ -1662,6 +1695,20 @@ bool TabBar::get_select_with_rmb() const {
+	}
+	deselect_enabled = p_enabled;
+	if (!deselect_enabled && !tabs.is_empty()) {
+		select_next_available();
```
This should only happen if current tab is -1.

kitbdev commented:
> The logic for setting -1 can be improved. Pseudocode:

This would change the logic. All tabs must be hidden or disabled for deselection, not any.
I can add the condition and change it to a for each loop though.

KoBeWi commented:
Right, I updated it.

kitbdev commented:
Fixed issue when removing all but a disabled tab would deselect that tab instead of none.
Fixed multiple tab_changed signals when moving tab.

KoBeWi commented on scene/gui/tab_bar.cpp:
```diff
@@ -1516,10 +1531,28 @@ void TabBar::_ensure_no_over_offset() {
+	bool can_deselect = true;
+	for (const Tab &tab : tabs) {
+		if (!tab.disabled && !tab.hidden) {
+			can_deselect = false;
```
Technically you can return `false` here and return `true` after the loop.

KoBeWi reviewed (approved):
I checked the linked issues and they aren't really reproducible, so they could be closed even without this PR I think.

kitbdev commented:
I added an MRP to https://github.com/godotengine/godot/issues/74653 with steps.
https://github.com/godotengine/godot/issues/73785 is a duplicate of it.

KoBeWi commented:
So your PR adds tab changing based on visibility 🤔
Well, can't say if it's expected/desirable, but at least it works correctly.

YeldhamDev commented:
When no tab is selected, the panel itself isn't draw. I don't think this should be how it works.

kitbdev commented:
> When no tab is selected, the panel itself isn't draw. This shouldn't be how it works.

This is intended, because I am planning to make the Editor's Bottom Dock into a TabContainer, and the Panel will need to hide.
I could add an option for it if we want?

YeldhamDev commented:
Then the "panel" theme style should be overridden with an empty one instead.

YuriSizov commented:
Why would we want to draw the panel if there is no content to display?

YeldhamDev commented:
- It's already the default behavior to always show the panel, even if there's no tabs available.
- It's visually displeasing to have the panel completely vanish while the tabs are still hanging in there.
- The panel can still be used to display some sort of content by the user, depending on their needs.

YuriSizov commented on scene/gui/tab_container.cpp:
```diff
@@ -460,6 +465,43 @@ void TabContainer::_on_active_tab_rearranged(int p_tab) {
+		return;
+	}
+	// Only allow one tab to be visible.
+	bool made_visibile = p_child->is_visible();
```
Typo :)

YuriSizov commented:
> * It's already the default behavior to always show the panel, even if there's no tabs available.
> * It's visually displeasing to have the panel completely vanish while the tabs are still hanging in there.

This is rather arbitrary. The default behavior doesn't change anyway, and if you enable the feature to allow closing all the tabs, then you want to be able to close all tabs, not just hide their content.

> * The panel can still be used to display some sort of content by the user, depending on their needs.

This is hard to control, so it would be limited in its usefulness. However, it got me thinking. We have `use_hidden_tabs_for_min_size`, which is not respected in this case. But it should. We should still draw the panel at least if we enable this option.

For the rest I'm not convinced it's a blocking issue.

YuriSizov reviewed (approved):
I think this is overall good, from reading the code.

But the question above is up in the air.

kitbdev commented:
Since I'm not using `use_hidden_tabs_for_min_size`, the panel will collapse away for me, and if it was then I could set the panel style to empty, so drawing the panel works for me.
